openhtf/util/measurements.py

```diff
@@ -326,18 +357,25 @@ def __getattr__(self, name):  # pylint: disable=invalid-name
 
   def __setitem__(self, name, value):  # pylint: disable=invalid-name
     self._AssertValidKey(name)
-    record = self._values.setdefault(
-        name, MeasuredValue.ForMeasurement(self._measurements[name]))
-    record.SetValue(value)
+    if self._measurements[name].dimensions:
+      raise InvalidDimensionsError(
+          'Cannot set dimensioned measurement without indices')
+    self._values.setdefault(
+        name, self._measurements[name].MakeUnsetValue()).value = value
 
   def __getitem__(self, name):  # pylint: disable=invalid-name
     self._AssertValidKey(name)
+
+    # __getitem__ is used to set dimensioned values via __setitem__ on the
+    # DimensionedMeasuredValue object, so we can't do any checking here.
     if self._measurements[name].dimensions:
-      return self._values.setdefault(name, MeasuredValue.ForMeasurement(
-          self._measurements[name]))
+      return self._values.setdefault(
```

Comment:
nit: using `setdefault` means making an unset value for every call, whether the value was set before or not. This can be costly for repeated accesses, especially with multidimensional measurements we expect repeated accesses for.